### ISSUE TITLE
Datatype not correctly parsed for Const-prop pass

### ISSUE THREAD

**Issue by hamptonm1** [onnx/onnx-mlir#2344]
When compiling a customer model, we see the following errors:

```
root@7bc1ee526dda:~/onnx-mlir/build# onnx-mlir --O3 --EmitLib --maccel=NNPA model.onnx --onnx-op-stats TXT -v
The ONNX model has 25071 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
loc("StatefulPartitionedCall/sequential/dense/MatMul"): error: reduction dimension must be the same
loc("StatefulPartitionedCall/sequential/dense/MatMul"): error: Failed to scan parameters successfully
loc("StatefulPartitionedCall/sequential/dense/MatMul"): error: 'onnx.MatMul' op shape inference failed
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Gemm 2nd dim of A is different than 1st dim of B
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Failed to scan parameters successfully
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: 'onnx.Gemm' op shape inference failed
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Gemm 2nd dim of A is different than 1st dim of B
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Failed to scan parameters successfully
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: 'onnx.Gemm' op shape inference failed
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Gemm 2nd dim of A is different than 1st dim of B
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Failed to scan parameters successfully
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: 'onnx.Gemm' op shape inference failed
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Gemm 2nd dim of A is different than 1st dim of B
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Failed to scan parameters successfully
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: 'onnx.Gemm' op shape inference failed
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Gemm 2nd dim of A is different than 1st dim of B
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Failed to scan parameters successfully
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: 'onnx.Gemm' op shape inference failed
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Gemm 2nd dim of A is different than 1st dim of B
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Failed to scan parameters successfully
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: 'onnx.Gemm' op shape inference failed
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Gemm 2nd dim of A is different than 1st dim of B
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: Failed to scan parameters successfully
loc(fused["StatefulPartitionedCall/sequential/dense/BiasAdd", "StatefulPartitionedCall/sequential/dense/MatMul"]): error: 'onnx.Gemm' op shape inference failed
onnx-mlir: /workdir/onnx-mlir/src/Dialect/Mlir/IndexExprBuilder.cpp:84: void onnx_mlir::IndexExprBuilder::assertHasShapeAndRank(mlir::Value): Assertion `hasShapeAndRank(value) && "expected value with shape and rank"' failed.
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted
```

**Comment by hamptonm1** [comment]
When I started debugging.... I went down several passes. The one pass that failed is `constprop-onnx` with the following error:

```
root@7bc1ee526dda:~/onnx-mlir/build# onnx-mlir-opt --maccel=NNPA model.onnx.mlir --constprop-onnx -v

ElementsAttr does not provide iteration facilities for type `float`, see attribute: dense_resource<__elided__> : tensor<1x64x1xf32>
invalid `T` for ElementsAttr::getValues
UNREACHABLE executed at /workdir/llvm-project/mlir/include/mlir/IR/BuiltinAttributeInterfaces.h:308!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --constprop-onnx -v
 #0 0x0000004004e303b0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4e303b0)
 #1 0x0000004004e2dc84 SignalHandler(int) Signals.cpp:0:0
 #2 0x00000040092f3520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #3 0x0000004009347a7c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #4 0x0000004009347a7c __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #5 0x0000004009347a7c pthread_kill ./nptl/pthread_kill.c:89:10
 #6 0x00000040092f3476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #7 0x00000040092d97f3 abort ./stdlib/abort.c:81:7
 #8 0x0000004004d7409e (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4d7409e)
 #9 0x0000004001d71030 _ZNK4mlir12ElementsAttr11value_beginIfEENSt9enable_ifIXoosrSt7is_sameINS_9AttributeET_E5valuentsrSt10is_base_ofIS4_S5_E5valueENS_6detail20ElementsAttrIteratorIS5_EEE4typeEv /workdir/llvm-project/mlir/include/mlir/IR/BuiltinAttributeInterfaces.h:304:31
#10 0x0000004001d6c157 _ZNK4mlir12ElementsAttr9getValuesIfEENSt9enable_ifIXoosrSt7is_sameINS_9AttributeET_E5valuentsrSt10is_base_ofIS4_S5_E5valueENS_6detail17ElementsAttrRangeINS9_20ElementsAttrIteratorIS5_EEEEE4typeEv /workdir/llvm-project/build/tools/mlir/include/mlir/IR/BuiltinAttributeInterfaces.h.inc:182:64
#11 0x0000004001d427ef auto onnx_mlir::ElementsAttrBuilder::allEqual(mlir::ElementsAttr, onnx_mlir::WideNum)::'lambda'(auto)::operator()<std::integral_constant<onnx_mlir::BType, (onnx_mlir::BType)1>>(auto) const /workdir/onnx-mlir/src/Dialect/ONNX/ElementsAttr/ElementsAttrBuilder.cpp:169:30
#12 0x0000004001d43b47 auto onnx_mlir::dispatchByBType<onnx_mlir::ElementsAttrBuilder::allEqual(mlir::ElementsAttr, onnx_mlir::WideNum)::'lambda'(auto)>(onnx_mlir::BType, auto&&) /workdir/onnx-mlir/src/Dialect/ONNX/ElementsAttr/BType.hpp:257:39
#13 0x0000004001d43c33 onnx_mlir::ElementsAttrBuilder::allEqual(mlir::ElementsAttr, onnx_mlir::WideNum) /workdir/onnx-mlir/src/Dialect/ONNX/ElementsAttr/ElementsAttrBuilder.cpp:172:1
#14 0x0000004000f4ea27 (anonymous namespace)::isConstOf(mlir::Value, double) /workdir/onnx-mlir/src/Transform/ONNX/ConstProp.cpp:131:1
#15 0x0000004000f58a53 (anonymous namespace)::AddZerosOnRhs::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /workdir/onnx-mlir/build/src/Transform/ONNX/ONNXConstProp.inc:955:9
#16 0x00000040029e59a8 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x29e59a8)
#17 0x00000040029bf5ff (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
#18 0x00000040029c42e4 mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x29c42e4)
#19 0x0000004000fc0c6d mlir::applyPatternsAndFoldGreedily(mlir::Operation*, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) /workdir/llvm-project/mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:115:43
#20 0x0000004000f75214 (anonymous namespace)::ConstPropONNXToONNXPass::runOnOperation() /workdir/onnx-mlir/src/Transform/ONNX/ConstProp.cpp:1015:13
#21 0x0000004004a8b589 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4a8b589)
#22 0x0000004004a8ba71 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4a8ba71)
#23 0x0000004004a8c446 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::'lambda'(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&)::operator()(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&) const Pass.cpp:0:0
#24 0x0000004004a8ac15 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4a8ac15)
#25 0x0000004004a8b315 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4a8b315)
#26 0x0000004004a8ba71 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4a8ba71)
#27 0x0000004004a8c2d4 mlir::PassManager::run(mlir::Operation*) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4a8c2d4)
#28 0x00000040011c3439 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#29 0x00000040011c459c processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPool*) MlirOptMain.cpp:0:0
#30 0x00000040011c46fd mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#31 0x0000004004d0a3b4 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x4d0a3b4)
#32 0x00000040011be1b4 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x11be1b4)
#33 0x0000004000401752 main /workdir/onnx-mlir/src/Tools/onnx-mlir-opt/onnx-mlir-opt.cpp:216:13
#34 0x00000040092dad90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#35 0x00000040092dae40 call_init ./csu/../csu/libc-start.c:128:20
#36 0x00000040092dae40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#37 0x00000040003ff625 _start (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x3ff625)
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted
```

**Comment by hamptonm1** [comment]
From my debugging it appears that the code is not correctly differentiating between float type vs. integer type. @sorenlassen said he will take a look at this specifically in `/workdir/onnx-mlir/src/Dialect/ONNX/ElementsAttr/ElementsAttrBuilder.cpp` lines 161-167

```
if (lhs.isSplat()) {
          cpptype x = lhs.getSplatValue<cpptype>();
          return nEquals(x);
        } else {
          auto values = lhs.getValues<cpptype>();
          return llvm::all_of(values, nEquals);
        }
```

**Comment by sorenlassen** [comment]
> I can't reproduce the problem locally but can you please try to build onnx-mlir with my fix from PR #2343 and see if it fixes the issue?

update: I found out how to recreate locally, and there's a bug in PR #2343 so don't try it yet

**Comment by hamptonm1** [comment]
> > I can't reproduce the problem locally but can you please try to build onnx-mlir with my fix from PR #2343 and see if it fixes the issue?
> 
> update: I found out how to recreate locally, and there's a bug in PR #2343 so don't try it yet

Okay... is it safe to try now? I pulled in your latest code updates in `ElementsAttrBuilder.cpp` to give it a try with the model.

**Comment by sorenlassen** [comment]
> is it safe to try now?

yes

**Comment by sorenlassen** [comment]
reopening this issue until we confirm that it was fixed

**Comment by sorenlassen** [comment]
I now notice in the second error posted in the issue that the ElementsAttr was a DenseResourceElementsAttr, like in issue #2346

I can write a quick patch for ElementsAttrBuilder::allEqual() but the model is likely to break somewhere else in constant propagation, because DenseResourceElementsAttr is generally unsupported

@tungld you mentioned in PR #2347 that it's worth supporting DenseResourceElementsAttr because it is an MLIR attribute and other dialects (perhaps in the future) will use it for normal/unencrypted data, but I don't think it's practical support it the way we use it in ZHigh now because take a DenseUI8ResourceElementsAttr with a flat shape and element type i8 (not even ui8) and set that as the value of a ZHighStickifiedConstantOp with different shape and element type: this makes it impossible to understand the structure of the data by looking at the ElementsAttr

it wasn't 100% clear to me from the stack trace how we ended up calling onnx constant propagation with DenseUI8ResourceElementsAttr but I'm guessing it's because addONNXToZHighPasses() adds ConstPropONNXToONNXPass - I think we need to remove that until we have a better solution to how we use DenseUI8ResourceElementsAttr

I think we can solve it by replacing DenseUI8ResourceElementsAttr with DisposableElementsAttr which has the capabilities to hold its data in a user-allocated buffer with custom alignment. This will require that, in the case of lowering ONNX to ZHigh, we delay ScrubDisposablePass until after ConstPropONNXToONNXPass in addONNXToZHighPasses().

(I'm also trying to make DisposableElementsAttr work throughout all the dialects and compiler passes and get rid of ScrubDisposablePass altogether but that will take a bit longer.)

**Comment by sorenlassen** [comment]
> I think we can solve it by replacing DenseUI8ResourceElementsAttr with DisposableElementsAttr which has the capabilities to hold its data in a user-allocated buffer with custom alignment. This will require that, in the case of lowering ONNX to ZHigh, we delay ScrubDisposablePass until after ConstPropONNXToONNXPass in addONNXToZHighPasses().

@tungld I took a closer look at how ZHigh uses DenseUI8ResourceElementsAttr as the value attribute in ZHighStickifiedConstantOp and I don't see any benefits over using a plain DenseElementsAttr. How about that we replace DenseUI8ResourceElementsAttr with DenseElementsAttr? This will simplify things in a few places, e.g. KrnlGlobalOp will always have a DenseElementsAttr.

**Comment by tungld** [comment]
@sorenlassen it's good to have a single DenseElementsAttr for all. Stickified data is dfloat16 which is not supported in MLIR or LLVM, that's why we have used DenseUI8ResourceElementsAttr (before ResourceElementsAttr, Opaque attribute is used for encoded data) . In such case, how do you use DenseElementsAttr?

**Comment by tungld** [comment]
@hamptonm1 as you know `onnx.Reshape` is an CPU  op, so every tensor passed to it must be not `a stickified tensor` (that is dedicated to NNPA). Could you find out why the input of `onnx.Reshape` was a stickified tensor?

**Comment by sorenlassen** [comment]
> @sorenlassen it's good to have a single DenseElementsAttr for all. Stickified data is dfloat16 which is not supported in MLIR or LLVM, that's why we have used DenseUI8ResourceElementsAttr (before ResourceElementsAttr, Opaque attribute is used for encoded data) . In such case, how do you use DenseElementsAttr?

I'm curious, is dfloat16 the DLFloat16 data type from the 2019 paper https://doi.org/10.1109/ARITH.2019.00023 ?

I think you could use a DenseElementsAttr with ui16 elements to represent the dfloat16 bits - it won't perfectly model the data but I think it will work just as well as DenseUI8ResourceElementsAttr and we will avoid the complications that comes with DenseResourceElementsAttr

**Comment by sorenlassen** [comment]
oops, again I accidentally closed the issue

reopening

@hamptonm1 please confirm if the problem still occurs after the fix in PR #2349

**Comment by sorenlassen** [comment]
> @hamptonm1 as you know `onnx.Reshape` is a CPU op, so every tensor passed to it must not be `a stickified tensor` (that is dedicated to NNPA). Could you find out why the input of `onnx.Reshape` was a stickified tensor?

we could use PR #2355 to try to catch how a DenseResourceElementsAttr from a stickified tensor ended up in a constant operand to onnx.Reshape

**Comment by hamptonm1** [comment]
I am so sorry for not responding sooner... I was fighting with getting Podman resolved. My apologies. I am really sorry @sorenlassen @tungld

**Comment by hamptonm1** [comment]
> @hamptonm1 as you know `onnx.Reshape` is a CPU op, so every tensor passed to it must not be `a stickified tensor` (that is dedicated to NNPA). Could you find out why the input of `onnx.Reshape` was a stickified tensor?

@tungld Yes I can do so.

**Comment by hamptonm1** [comment]
@sorenlassen I noticed the PR is merged.... I directly sent you my tested results but I will post it here just in case too. 

```
root@3292372007da:~/onnx-mlir/build# onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --decompose-onnx >& decompose-onnx
root@3292372007da:~/onnx-mlir/build# vim decompose-onnx 
root@3292372007da:~/onnx-mlir/build# onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --onnx-hybrid-transform >& onnx-hybrid-transform
Segmentation fault (core dumped)
root@3292372007da:~/onnx-mlir/build# vim onnx-hybrid-transform 
root@3292372007da:~/onnx-mlir/build#  onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --conv-opt-onnx >&conv-opt-onnx
root@3292372007da:~/onnx-mlir/build# vim conv-opt-onnx 
root@3292372007da:~/onnx-mlir/build# onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --shape-inference >& shape-inference
Segmentation fault (core dumped)
root@3292372007da:~/onnx-mlir/build# vim shape-inference 





[2:11](https://lfaifoundation.slack.com/archives/D05162XMTK6/p1689012688137489)
onnx-hybrid-transform

ElementsAttr does not provide iteration facilities for type `llvm::APInt`, see attribute: dense_resource<__elided__> : tensor<4xi64>
invalid `T` for ElementsAttr::getValues
UNREACHABLE executed at /workdir/llvm-project/mlir/include/mlir/IR/BuiltinAttributeInterfaces.h:308!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --onnx-hybrid-transform
~                                                                                                                                                     
~                                                                                                                                                     
~                                                                                                                                                     
~          
[2:12](https://lfaifoundation.slack.com/archives/D05162XMTK6/p1689012733627979)
shape-inference

ElementsAttr does not provide iteration facilities for type `llvm::APInt`, see attribute: dense_resource<__elided__> : tensor<4xi64>
invalid `T` for ElementsAttr::getValues
UNREACHABLE executed at /workdir/llvm-project/mlir/include/mlir/IR/BuiltinAttributeInterfaces.h:308!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --shape-inference
[2:13](https://lfaifoundation.slack.com/archives/D05162XMTK6/p1689012817060799)
constprop-onnx

onnx-mlir-opt: /workdir/onnx-mlir/src/Dialect/ONNX/ONNXOps/OpHelper.cpp:515: bool onnx_mlir::isDenseONNXConstant(mlir::Value): Assertion `(isa<DenseElementsAttr, DisposableElementsAttr>(value)) && "unsupported onnx constant value attribute"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt --maccel=NNPA fraud_cnn_model.onnx.mlir --constprop-onnx
~                                                                                                                                                     
~
```

**Comment by hamptonm1** [comment]
The three passes that are still failing are `onnx-hybrid-transform`, `shape-inference`, and `constprop-onnx`. However, the error messages are just different at this point.

**Comment by sorenlassen** [comment]
the `OpHelper.cpp:515` assert was introduced in PR #2355 to troubleshoot why DenseResourceElementsAttr shows up in ONNXConstantOps

if we can get the stack traces of all the failures, hopefully we can figure out where they come from

**Comment by hamptonm1** [comment]
@sorenlassen So good news.... this issue has nothing to do with your code but instead the model. I worked with @tungld last night and it seems like for `onnx.reshape` the input tensor value is not correct which is right before `onnx.matmul` operation. When we change the input value of the tensor to the correct value ( which should be an unstickfied tensor) in the .`onnx.mlir` file and then run through all three passes that were previously failing....it now works perfectly fine. Sorry for all of the confusion but thank you for being so willing to help. I am glad we got down to the root of the problem. Using Netron help to visualize everything and thank you @tungld for teaching me :)
